src/dune_engine/load_rules.ml

```diff
@@ -279,7 +279,10 @@ let no_rule_found ~loc fn =
     Code_error.raise ?loc "Build_system.no_rule_found got anonymous action path"
       [ ("fn", Path.Build.to_dyn fn) ]
 
+(* XXX This function is confusingly named. It works on external paths just as
+   well *)
 let source_file_digest path =
+  assert (not (Path.is_in_build_dir path));
```

Comment:
This assertion isn't quite right, because it makes the following test fail:

```
 |# CR-someday amokhov: Apparently, Dune thinks that [../link] points to the build
 |# directory and raises a code error. Fix this.
 |
 |  $ cat ../link
 |  hi
 |  $ dune build t --sandbox hardlink 2>&1
-|  Internal error, please report upstream including the contents of _build/log.
-|  Description:
-|    ("Fs_memo.Watcher.watch called on a build path",
-|    { path = In_build_dir "link" })
+|  File "src/dune_engine/load_rules.ml", line 285, characters 2-8:
+|  File "src/dune_engine/load_rules.ml", line 285, characters 2-8: Assertion
+|  failed
+|  Raised at Dune_engine__Load_rules.source_file_digest in file
+|    "src/dune_engine/load_rules.ml", line 285, characters 2-42
+|  Called from Dune_engine__Load_rules.get_rule_or_source in file
+|    "src/dune_engine/load_rules.ml", line 881, characters 13-36
+|  Called from Dune_engine__Build_system.Exported.build_file_impl in file
+|    "src/dune_engine/build_system.ml", line 888, characters 4-38
+|  Called from Fiber__Scheduler.exec in file "src/fiber/scheduler.ml", line 69,
+|    characters 8-11
```

Reply:
I'm not sure what you mean: the error above says it's coming from `Fs_memo.Watcher.watch`, not from `source_file_digest`.

Reply:
Oh, it was coming from `Fs_memo.Watcher.watch` and now it's coming from here.

Anyway, I think this assertion is trying to check a useful fact. We don't want to use it on files in the build directory.

Reply:
Renamed.